File: testsuite/tests/ada/0006-strings/cpp/main.cpp
#include <iostream>

#include <polyglot_ada_strings.h>
#include "test.h"

using namespace polyglot::ada::strings;

int main() {
    polyglot_string ada_arr = test::string_func();
    test::string_proc(ada_arr);
    std::string ada_str = to_string(ada_arr);
    std::cout << "Ada.String from C++: " << ada_str << std::endl;

    polyglot_string user_arr = test::user_string_func();
    test::user_string_proc(user_arr);
    std::string user_str = to_string(user_arr);
    std::cout << "Test.User_Str from C++: " << user_str << std::endl;

    // We should be able to use functions defined for similar types
    test::user_string_proc(ada_arr);
    test::string_proc(user_arr);

    // Test with a string from C++
    std::string cpp_str = "from C++";
    polyglot_string cpp_arr = from_string(cpp_str);
    test::string_proc(cpp_arr);
    test::user_string_proc(cpp_arr);
}

File: testsuite/tests/ada/0006-strings/rust/main.rs
use test::test as pkg;
use test::ada::strings::PolyglotString;

fn main() {
    let ada_arr = pkg::string_func();
    pkg::string_proc(&ada_arr);
    println!("Ada.String from Rust: {}", ada_arr);

    let user_arr = pkg::user_string_func();
    pkg::user_string_proc(&user_arr);
    println!("Test.User_Str from Rust: {}", user_arr);

    // We should be able to use functions defined for similar types.
    pkg::user_string_proc(&ada_arr);
    pkg::string_proc(&user_arr);

    // A string built on the Rust side can be passed across the boundary too.
    let rust_arr = PolyglotString::from("from Rust");
    pkg::string_proc(&rust_arr);
    pkg::user_string_proc(&rust_arr);
}
